File: core/concurrent/java/fan/concurrent/AtomicRefPeer.java
//
// Copyright (c) 2010, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   12 Feb 10 Brian Frank  Creation
//
package fan.concurrent;

import java.util.concurrent.atomic.AtomicReference;

import fan.sys.*;

public final class AtomicRefPeer
{

  public static AtomicRefPeer make(AtomicRef self)
  {
    return new AtomicRefPeer();
  }

  public final Object val(AtomicRef self)
  {
    return val.get();
  }

  public final void val(AtomicRef self, Object v)
  {
    if (!FanObj.isImmutable(v)) throw NotImmutableErr.make();
    val.set(v);
  }

  public final Object getAndSet(AtomicRef self, Object v)
  {
    if (!FanObj.isImmutable(v)) throw NotImmutableErr.make();
    return val.getAndSet(v);
  }

  public final boolean compareAndSet(AtomicRef self, Object expect, Object update)
  {
    if (!FanObj.isImmutable(update)) throw NotImmutableErr.make();
    return val.compareAndSet(expect, update);
  }

  private final AtomicReference val = new AtomicReference();
}

File: src/concurrent/js/AtomicRefPeer.js
//
// Copyright (c) 2012, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   2 Nov 2012  Andy Frank  Creation
//

/**
 * AtomicRefPeer.
 */
fan.concurrent.AtomicRefPeer = fan.sys.Obj.$extend(fan.sys.Obj);
fan.concurrent.AtomicRefPeer.prototype.$ctor = function() {}

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

fan.concurrent.AtomicRefPeer.prototype.m_val = null;
fan.concurrent.AtomicRefPeer.prototype.val = function(self) { return this.m_val; }
fan.concurrent.AtomicRefPeer.prototype.val$ = function(self, val)
{
  if (!fan.sys.ObjUtil.isImmutable(val)) throw fan.sys.NotImmutableErr.make();
  this.m_val = val;
}

fan.concurrent.AtomicRefPeer.prototype.getAndSet = function(self, val)
{
  if (!fan.sys.ObjUtil.isImmutable(val)) throw fan.sys.NotImmutableErr.make();
  var old = this.m_val;
  this.m_val = val;
  return old;
}

fan.concurrent.AtomicRefPeer.prototype.compareAndSet = function(self, expect, update)
{
  if (!fan.sys.ObjUtil.isImmutable(update)) throw fan.sys.NotImmutableErr.make();
  if (this.m_val != expect) return false;
  this.m_val = update;
  return true;
}
